src/workerd/server/server.c++

```diff
@@ -904,6 +905,172 @@ kj::Own<Server::Service> Server::makeDiskDirectoryService(
+class Server::InspectorService final: public kj::HttpService, public kj::HttpServerErrorHandler {
+  // Implements the interface for the devtools inspector protocol.
+  //
+  // The InspectorService is created when workerd serve is called using the -i option
+  // to define the inspector socket.
+public:
+  InspectorService(
+      kj::Timer& timer,
+      kj::HttpHeaderTable::Builder& headerTableBuilder)
+      : timer(timer),
+        headerTable(headerTableBuilder.getFutureTable()),
+        server(timer, headerTable, *this, kj::HttpServerSettings {
+          .errorHandler = *this
+        }) {}
+
+  kj::Promise<void> handleApplicationError(
+      kj::Exception exception, kj::Maybe<kj::HttpService::Response&> response) override {
+    KJ_LOG(ERROR, kj::str("Uncaught exception: ", exception));
+    KJ_IF_MAYBE(r, response) {
+      return r->sendError(500, "Internal Server Error", headerTable);
+    } else {
+      return kj::READY_NOW;
+    }
+  }
+
+  kj::Promise<void> request(
+      kj::HttpMethod method,
+      kj::StringPtr url,
+      const kj::HttpHeaders& headers,
+      kj::AsyncInputStream& requestBody,
+      kj::HttpService::Response& response) override {
+    // The inspector protocol starts with the debug client sending ordinary HTTP GET requests
+    // to /json/version and then to /json or /json/list. These must respond with valid JSON
+    // documents that list the details of what isolates are available for inspection. Each
+    // isolate must be listed separately. In the advertisement for each isolate is a URL
+    // and a unique ID. The client will use the URL and ID to open a WebSocket request to
+    // actually connect the debug session.
+    kj::HttpHeaders responseHeaders(headerTable);
+    if (headers.isWebSocket()) {
+      KJ_IF_MAYBE(pos, url.findLast('/')) {
+        auto id = url.slice(*pos + 1);
+
+        KJ_IF_MAYBE(isolate, isolates.find(id)) {
+          // If getting the strong ref doesn't work it means that the Worker::Isolate
+          // has already been cleaned up. We use a weak ref here in order to keep from
+          // having the Worker::Isolate itself having to know anything at all about the
+          // IsolateService and the registration process. So instead of having Isolate
+          // explicitly clean up after itself we lazily evaluate the weak ref and clean
+          // up when necessary.
+          KJ_IF_MAYBE(ref, (*isolate)->tryAddStrongRef()) {
+            // When using --verbose, we'll output some logging to indicate when the
+            // inspector client is attached/detached.
+            KJ_LOG(INFO, kj::str("Inspector client attaching [", id, "]"));
+            auto webSocket = response.acceptWebSocket(responseHeaders);
+            kj::Duration timerOffset = 0 * kj::MILLISECONDS;
+            return (*ref)->attachInspector(timer, timerOffset, *webSocket)
+                .attach(kj::mv(webSocket)).catch_([id=kj::str(id)](kj::Exception&& ex)
```

Comment:
Hmm it looks like you didn't implement attaching the isolate ref to the `attachInspector()` promise? I think this will do it:

```suggestion
                .attach(kj::mv(webSocket), kj::mv(*ref))
                .catch_([id=kj::str(id)](kj::Exception&& ex)
```